### ISSUE TITLE
OpenBSD support?

### ISSUE THREAD

**Issue by neezer** [tonarino/innernet#25]
Any chance for this? I see `shared/src/wg.rs` only targets `macos` & `linux`... thanks!!

**Comment by mcginty** [comment]
Would love to support OpenBSD - we accept PRs!

**Comment by mrdomino** [comment]
If I’m not mistaken (and I might be mistaken), it seems like on OpenBSD you want to use [`wg(4)`](https://man.openbsd.org/wg.4). So you would create a `wgN` interface using (e.g.) `ifconfig wg0 create`. Then, everything looks to be implemented with a couple `ioctl(2)`s: `SIOCSWG struct wg_data_io*` to set device configuration, and `SIOCGWG struct wg_data_io*` to get it.

It seems then like there are two possible approaches: one is to basically shell out to `ifconfig(8)` on OpenBSD, and the other is to reimplement the `wg`-specific functionality from `ifconfig(8)` in `wgctrl-sys`. I don’t have any principled opinions on which is better; if one wanted to do the latter, then [the `ifconfig(8)` source](https://github.com/openbsd/src/blob/master/sbin/ifconfig/ifconfig.c#L5708) might be a good starting point.

**Comment by mrdomino** [comment]
(I would also love to see innernet support OpenBSD.)

**Comment by mrdomino** [comment]
Also minor note for anyone getting started on this: the first time I tried to build `wgctrl-rs`, I saw:

```
thread 'main' panicked at 'Unable to find libclang: "couldn't find any of ['libclang.so', 'libclang.so.*', 'libclang-*.so'], set the LIBCLANG_PATH environment variable to a path where one of these files can be found (skipped: [])"', /home/burld/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.39.0/src/lib.rs:1605:31
```

This was easily fixed by doing `pkg_add llvm` and setting `LIBCLANG_PATH=/usr/local/lib`.

**Comment by mcginty** [comment]
@mrdomino yep, things have moved around a bit (`wgctrl-rs` is now `wireguard-control`, and there is no more `wgctrl-sys` since we use the `netlink` crate now instead of an FFI), but the basic need for supporting OpenBSD is adding an `wireguard-control/src/backends/openbsd.rs` that can communicate with the pseudo-device with the correct `ioctl`s (ideally) or shell out to `ifconfig` (less ideally since I've found that to be quite brittle, but that's how we do it on macOS currently too).

**Comment by alpn** [comment]
PR #347 is a proposed implementation.

I experimented with both approaches ([`ifconfig(8)`](https://man.openbsd.org/ifconfig) and [`ioctl(2)`](https://man.openbsd.org/ioctl)) and ultimately decided to go with the former, as it's simpler and avoids introducing new `unsafe` blocks.

Parsing was implemented directly against [ifconfig.c](https://github.com/openbsd/src/blob/master/sbin/ifconfig/ifconfig.c).
Any deviation from the expected format is considered a violation of the interface.

Tested on `7.7` and `current`.
I've been running it for over a month with no issues.

@mcginty Still interested in reviewing?

**Comment by mcginty** [comment]
@alpn I'm sorry to disappoint, but I don't contribute to tonari any more, and have transferred ownership to their current engineers, so it's in their hands to review large PRs. Glad to see the support being added though, best of luck.

**Comment by strohel** [comment]
Sorry for the wait @alpn and others, reviewed!

FWIW @mrdomino (and everyone): it's still helpful to review and discuss PRs that you're interested in even if you're not formally a maintainer. 👌

**Comment by mrdomino** [comment]
Yeah I mean, overall it looked fine to me; I couldn't tell how normal all of the `expect` fails were vs returning errors, and it's been a while since I've had OpenBSD networking state paged in, but otherwise no complaints.
